pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package migration
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/google/uuid"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	vim25types "github.com/vmware/govmomi/vim25/types"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	migrationv1alpha1 "sigs.k8s.io/vsphere-csi-driver/pkg/apis/migration/v1alpha1"
+	cnsvolume "sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/volume"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	cnsconfig "sigs.k8s.io/vsphere-csi-driver/pkg/common/config"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+// VolumeMigrationService exposes interfaces to support VCP to CSI migration.
+// It will maintain internal state to map volume path to volume ID and reverse mapping.
+type VolumeMigrationService interface {
+	// GetVolumeID returns VolumeID for given VolumePath
+	// Returns an error if not able to retrieve VolumeID.
+	GetVolumeID(ctx context.Context, volumePath string) (string, error)
+
+	// GetVolumePath returns VolumePath for given VolumeID
+	// Returns an error if not able to retrieve VolumePath.
+	GetVolumePath(ctx context.Context, volumeID string) (string, error)
+
+	// DeleteVolumeInfo helps delete mapping of volumePath to VolumeID for specified volumeID
+	DeleteVolumeInfo(ctx context.Context, volumeID string) error
+}
+
+// volumeMigration holds migrated volume information and provides functionality around it.
+type volumeMigration struct {
+	// volumePath to volumeId map
+	volumePathToVolumeID sync.Map
+	// k8sClient helps operate on CnsVSphereVolumeMigration custom resource
+	k8sClient client.Client
+	// volumeManager helps connecting to vCenter to perform Volume Operations
+	volumeManager *cnsvolume.Manager
+	cnsConfig     *cnsconfig.Config
```

Comment:
Can you add a comment here why cnsConfig is used by this struct

Reply:
yes. adding as below

```
	// cnsConfig helps retrieve vSphere CSI configuration for RegisterVolume Operation
	cnsConfig     *cnsconfig.Config
```